lib/common/hash.go

```diff
@@ -1,18 +1,140 @@
+//
+// This package defines hash-related functions used in Sebak
+// The primary hash function used in Sebak is a double sha256,
+// like Bitcoin.
+//
 package common
 
 import (
+	"bytes"
 	"crypto/sha256"
+	"io"
+
 	"github.com/btcsuite/btcutil/base58"
 	ethcommon "github.com/ethereum/go-ethereum/common"
 	"github.com/ethereum/go-ethereum/rlp"
 )
 
+// 32 bytes / 256 bits hash type
+type Hash = ethcommon.Hash
+
+// Set the content of the hash to be the provided binary data
+var BytesToHash = ethcommon.BytesToHash
+
+// Generate a double SHA-256 hash out of the supplied binary data
 func MakeHash(b []byte) []byte {
 	first := sha256.Sum256(b)
 	second := sha256.Sum256(first[:])
 	return second[:]
 }
 
+// Interface recognized by RLP encoder
+type Encoder = rlp.Encoder
+
+// Encode the provided value
+// It is exposed here as it is useful for recursive calls
+// by types implementing `Encoder`
+var Encode = rlp.Encode
+
+// Ditto
+var EncodeToReader = rlp.EncodeToReader
+
+// Computes the minimum number of bytes required to store `i` in RLP encoding.
+func SizeofSize(i uint64) (size byte) {
+	for size = 1; ; size++ {
+		if i >>= 8; i == 0 {
+			return size
+		}
+	}
+}
+
+// Encodes a list to RLP, including its size
+func EncodeList(items ...interface{}) ([]byte, error) {
```

Comment:
The function you linked is private, so I cannot use it.
And it does not provide the same kind of functionality.

As I said in [this message](https://github.com/bosnet/sebak/pull/836#issuecomment-444562601), the header gets encoded as a list with a single element. This function allows to encode list, and no public function in the go-ethereum code lets me do that.

I am considering making a PR to go-ethereum to make some stuff available and remove a bunch of code from here, but I did not want to block this PR on another project.

Reply:
https://github.com/ethereum/go-ethereum/blob/3ac633ba8423851a85cb909a1f1f0f3c0d2c5351/rlp/encode.go#L80-L93
By using the function `func Encode(w io.Writer, val interface{}) error`,
We can encode an object according to its type.

But I understand your intention.
If there is a test suite that ensures the functions work well, it will be good.
How do you think about this?

Reply:
Indeed, but I want to encode the object without having to make an extra type.
I'll add tests for it.

Reply:
Why does `EncodeList` need?  As told @kfangw ,  By using `Encode(w,v) e` , we can encode any object even string or int. I am not sure but `EncodeList` has weak compatibility with outside sebak/go

Reply:
Calling `Encode` will only encode one object, while `EncodeList` encodes multiple object into a list.
So in pseudo-code:
```
var a, b, c
EncodeList(a, b, c) => [a, b, c]
Encode(a); Encode(b); Encode(c) => [a] [b] [c]
```

That's why I needed the extra function

Reply:
type xs []string
...
xs.Encode(...) 

It doesn’t works?

Reply:
```go
func TestAnarcher (t *testing.T) {
    if r1, e1 := EncodeList(uint(15), uint(9), uint(2)); e1 != nil {
        panic(e1)
    } else {
        fmt.Println(r1)
    }
    var buff bytes.Buffer
    w := bufio.NewWriter(&buff)
    Encode(w, uint(15))
    Encode(w, uint(9))
    Encode(w, uint(2))
    w.Flush()
    fmt.Println(buff.Bytes())
    t.FailNow()
}
```

```
[195 15 9 2]
[15 9 2]
```

The first one is a list, the second one just consecutive items.

Reply:
Actually there is a much simpler way to do this: https://github.com/bosnet/sebak/pull/903